frontend/catalyst/jax_tracer.py

```diff
@@ -1191,3 +1213,411 @@ def is_leaf(obj):
+                    qreg_out = qrp_out.actualize()
+
+                    meas_tracers = [trace.full_raise(m) for m in meas]
+                    meas_results = tree_unflatten(meas_trees, meas_tracers)
+
+                    # TODO: Allow the user to return whatever types they specify.
+                    if qnode_transformed or device_modify_measurements:
+                        assert isinstance(meas_results, list)
+                        if len(meas_results) == 1:
+                            transformed_results.append(meas_results[0])
+                        else:
+                            transformed_results.append(tuple(meas_results))
+                    else:
+                        transformed_results.append(meas_results)
+
+                    # Deallocate the register after each tape
+                    qdealloc_p.bind(qreg_out)
+
+                    # Get the jaxpr for this tape and clear the tracing frame for next tape
+                    # breakpoint()
+                    if type(transformed_results[i]) is not tuple:
+                        out_tracers = [transformed_results[i]]
+                    else:
+                        out_tracers = [_ for _ in transformed_results[i]]
+                    jp = EvaluationContext.find_jaxpr_frame().to_jaxpr(out_tracers)
+
+                    # EvaluationContext.find_jaxpr_frame().eqns.clear()
+                    EvaluationContext.find_jaxpr_frame().eqns.append(
+                        new_jaxpr_eqn([], [], jax.core.Primitive("_tape_cut"), [], no_effects)
+                    )
+                    # breakpoint()
+
+                    # get the eqn that should be the outvar for this tape's jaxpr
+                    # taken from
+                    # https://github.com/google/jax/blob/77afe251e70500d6e505449242b4343cb86c402f/jax/_src/interpreters/partial_eval.py#L1808
+                    invars = EvaluationContext.find_jaxpr_frame().invars
+                    frame = EvaluationContext.find_jaxpr_frame()
+                    state_ans, end_trees = unzip2(
+                        tree_flatten(t) for t in get_states(frame.attrs_tracked)
+                    )
+                    state_outvars = [
+                        frame.tracer_to_var[id(trace.full_raise(x))] for xs in state_ans for x in xs
+                    ]
+                    explicit_outvars = [frame.tracer_to_var[id(t)] for t in out_tracers]
+                    outvars = state_outvars + explicit_outvars
+                    constvars, constvals = unzip2(frame.constvar_to_val.items())
+                    jaxpr_effects = make_jaxpr_effects(
+                        constvars, invars, explicit_outvars, frame.eqns
+                    )
+
+                    jaxpr_info_each_tape.append(
+                        (
+                            [
+                                jax.core.Var("", _.aval) for _ in jp[0].invars
+                            ],  # maybe use frame here too?
```

Comment:
Also, this would make all tapes take the same input shape/type, since `jp` comes from the frame and the frame is shared across all tapes. Instead it would be nice to find a way to give each tape the inputs of that particular tape. 

Note that in the old implementation (with the reset) this was not an issue, as all tapes were chained into one big tape.